plugins/org.jkiss.dbeaver.model.ai/src/org/jkiss/dbeaver/model/ai/gpt3/GPTCompletionEngine.java

```diff
@@ -196,13 +196,20 @@ private String requestCompletion(
 
             try {
                 List<?> choices;
+                int responseSize = GPT_MODEL_MAX_RESPONSE_TOKENS;
                 for (int i = 0; ; i++) {
                     try {
                         choices = getCompletionChoices(service, completionRequest);
                         break;
                     } catch (Exception e) {
-                        if (e instanceof HttpException && ((HttpException) e).code() == 429) {
-                            RuntimeUtils.pause(1000);
+                        if ((e instanceof HttpException && ((HttpException) e).code() == 429)
+                            || (e instanceof OpenAiHttpException && e.getMessage().contains("This model's maximum"))) {
+                            if (e instanceof HttpException) {
+                                RuntimeUtils.pause(1000);
+                            } else {
+                                responseSize /= 2;
```

Comment:
It is not "smart" size detection. We can analyze error message and calculate exact maximum size of response.